README.md
# PostMessenger

`window.postMessage` is used to send messages between window objects on a page but receiving a response or acknowledgment is not built in. PostMessenger connects window objects and wraps `window.postMessage` messages in promises to make communication between windows easier to manage.

## Example Usage

Since this library is just a wrapper, PostMessenger should work between any two windows objects that postMessage supports. As an example we will assume you are trying to connect a root page window and an iframe loaded there. The easiest way to get started is to connect the two windows by instantiating PostMessenger in scripts on both windows. From the root page window:

```javascript
import { PostMessenger } from '@croia/post-messenger';

// Create and append the iframe to communicate with:
const iframeSrc = 'https://some.app';
const iframe = document.createElement('iframe');
iframe.src = iframeSrc;
document.body.appendChild(iframe);

const postMessenger = new PostMessenger({
  requestNames: {
    // These are just some example request names, call yours whatever you'd like.
    initializeIFrame: 'mainApp:initializeIFrame',
    requestDataFromRootWindow: 'mainApp:requestDataFromRootWindow',
    asyncRequestFromRootWindow: 'mainApp:asyncRequestFromRootWindow',
  },
  enableLogging: true,
  // Help identify this instance in the logs with a name:
  clientName: 'parent-client',
});

// Functions to run after receiving a request from the connected window with a matching requestName key
postMessenger.bindResponders({
  requestDataFromRootWindow: ({ someParam }) => {
    // You can return data back to the other window but it should be JSON serializable
    return someParam + 1;
  },
  // Return a promise and PostMessenger will await it for you and return any errors to the other window
  asyncRequestFromRootWindow: async ({ someValue }) => {
    const result = await makeAnAsyncRequest(someValue);
    return result;
  }
});

// Wait for connection from the iframe:
await postMessenger.acceptConnections({ origin: iframeSrc });

// Now that we are connected we can send and await requests to the iframe
const response = await postMessenger.request(postMessenger.requestNames.initializeIFrame, {
  // include any data you'd like but it will need to be JSON serializable:
  someMessage: '1234',
});

console.log({ response }); // { response: true } (iframe responder returns true, see example below)
```

From the iframe:

```javascript
import { PostMessenger } from '@croia/post-messenger';

const postMessenger = new PostMessenger({
  types: {
    initializeIFrame: 'mainApp:initializeTheApp',
    requestDataFromRootWindow: 'mainApp:requestDataFromRootWindow',
    asyncRequestFromRootWindow: 'mainApp:asyncRequestFromRootWindow',
  },
  enableLogging: true,
  clientName: 'child-client',
});

postMessenger.bindResponders({
  initializeIFrame: ({ someMessage }) => {
    console.log({ someMessage }); // { someMessage: '1234' }
    return true;
  }
});

// Initialize connection with the root window
await postMessenger.connect({
  targetOrigin: 'https://root-page.app',
  // Use window.parent here since this is an iframe and we are trying to connect to the parent page:
  targetWindow: window.parent,
});

// Now that we are connected we can send and await requests to the root window
const response = await postMessenger.request(postMessenger.requestNames.requestDataFromRootWindow, 3);

console.log({ response }); // { response: 4 }
```

## PostMessenger constructor options
`clientName` (string, optional), default `'unknown'`: A name for the PostMessenger instance that is useful for distinguishing clients in the console logs (if enabled).

`enableLogging` (boolean, optional), default `false`: If true will output logs when sending or receiving requests and the associated data.

`maxResponseTime` in milliseconds (number, optional), default `10000`: The max amount of time to wait before considering a request failed and rejecting the promise. Possible reasons for a timeout include the window being disconnected after the connection was established or an async request taking too long. Alternatively you can provide this as an option to an individual request call if you expect a specific request to take longer, e.g.

```javascript
const slowReq = await postMessenger.request(
  postMessenger.requestNames.slowReq,
  { maxResponseTime: 20000 },
);
```

`requestNames` ({ [string]: string }, required): Map where the values are the names of requests that are sent or received by the current window or the window you are connecting to. For example if you need to fetch some data in the iframe from the root page you might have a request name `fetchDataFromRootPage`. In general the `requestNames` option should be the exact same in both windows, containing the names of all requests sent between the two.

Providing the request names up front instead of sending and listening for arbtrary messages provides an advantage if you're using TypeScript. If so, PostMessenger will validate that the keys provided to the request and bindResponders functions exist on the `requestNames` provided when creating the PostMessenger instance. There are a couple other benefits to providing the names up front, such as validating the connected window is expecting the requestName and throwing an error immediately if not, and also allowing multiple PostMessenger instances to avoid request name collisions (e.g. by prefixing request names with "appOne:" or "appTwo:"). However these two issues could be resolved automatically in a future release by handling all requests through a common request wrapper unique to each instance, and requestNames would then be optional for additional type safety.

`useEncryption` (boolean, optional), default `true`: See Message Encryption section below for more details.

## PostMessenger.connect options

`targetOrigin` (string, required): The origin of the window to send the message to, either the URI or '*'. See also https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#targetorigin.

`targetWindow` (Window, required): Window to call postMessage on.

`maxRetries` (number, optional), default `10`: If the first connection attempt is unsuccessful (the window calling `connect` fails to receive a response from the specified window), PostMessenger will automatically retry once every 500 milliseconds up to the `maxRetries` value. This may help resolve minor race conditions when acceptConnections has not been set up yet in the root window.

## Security

In general be sure to follow the security recommendations [here](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#security_concerns). You should always specify both window origins when possible and PostMessenger will verify all messages are to and from the expected origin.
 
## Message encryption
Messages are encrypted by default which is useful when other potentially untrusted scripts are running on the same domain. For example if you're building an extension that needs to communicate from the root page to a trusted iframe you may not have control over all the scripts loaded by the root page. These scripts could intercept messages coming back from the trusted iframe to the root page domain. By default PostMessenger will generate an encryption key that is passed to the secondary trusted domain so it is known only to your script and the trusted domain. If this is not a concern for your app you can set `useEncryption: false`.

## acceptConnections

By default a specific root page `origin` must be provided `allowAnyOrigin` is `false` by default. If for some reason it's not possible to know in advance the domain of the root page that is sending the connection request you can pass `allowAnyOrigin: true` but be aware that any page could simulate a connection request to your app. You must take extra care in this case not to expose sensitive information or else ensure your app is only able to perform sensitive tasks using an API access key provided by the root page domain that a malicious third party wouldn't have.


CHANGELOG.md
# [2.0.0](https://github.com/croia/post-messenger/compare/v1.0.10...v2.0.0) (2022-11-07)


### Features

* Make requestNames optional and most methods private ([#62](https://github.com/croia/post-messenger/issues/62)) ([8df03f4](https://github.com/croia/post-messenger/commit/8df03f48e529a0be62d3381c7049f89c85a80a89))


### BREAKING CHANGES

* many previously available methods are now private

## [1.0.10](https://github.com/croia/post-messenger/compare/v1.0.9...v1.0.10) (2022-10-23)


### Bug Fixes

* lock file and add install to README ([#50](https://github.com/croia/post-messenger/issues/50)) ([880906f](https://github.com/croia/post-messenger/commit/880906fcfaf7d5b04423b5d8e42b4bd0396dc949))

## [1.0.9](https://github.com/croia/post-messenger/compare/v1.0.8...v1.0.9) (2022-10-23)


### Bug Fixes

* add package keywords ([#49](https://github.com/croia/post-messenger/issues/49)) ([31de5b4](https://github.com/croia/post-messenger/commit/31de5b47e39bc3fc2457359b158eed6497990c4f))

## [1.0.8](https://github.com/croia/post-messenger/compare/v1.0.7...v1.0.8) (2022-10-23)


### Bug Fixes

* copy README to package contents ([#48](https://github.com/croia/post-messenger/issues/48)) ([1767b7b](https://github.com/croia/post-messenger/commit/1767b7b38a2d41e7718305757eb65ecf30316571))

## [1.0.7](https://github.com/croia/post-messenger/compare/v1.0.6...v1.0.7) (2022-10-23)


### Bug Fixes

* npm access scope ([#47](https://github.com/croia/post-messenger/issues/47)) ([e3c15ec](https://github.com/croia/post-messenger/commit/e3c15ecb09da4c631b7dbf9074d4005893f9e610))

## [1.0.6](https://github.com/croia/post-messenger/compare/v1.0.5...v1.0.6) (2022-10-23)


### Bug Fixes

* update docs and trigger release ([#46](https://github.com/croia/post-messenger/issues/46)) ([b9250b1](https://github.com/croia/post-messenger/commit/b9250b16fad361f253948da4639d80690b07df6e))

## [1.0.5](https://github.com/croia/post-messenger/compare/v1.0.4...v1.0.5) (2022-10-23)


### Bug Fixes

* switch package location to registry.npmjs.org ([#45](https://github.com/croia/post-messenger/issues/45)) ([5236dc1](https://github.com/croia/post-messenger/commit/5236dc10f1896235464d7a91302e6964182a967c))

## [1.0.4](https://github.com/croia/post-messenger/compare/v1.0.3...v1.0.4) (2022-10-19)


### Bug Fixes

* change `types` to less ambiguous `requestNames` and improve docs ([#27](https://github.com/croia/post-messenger/issues/27)) ([a327d23](https://github.com/croia/post-messenger/commit/a327d23a38b29e59b898b6a0a32613f53c9ebbab))

## [1.0.3](https://github.com/croia/post-messenger/compare/v1.0.2...v1.0.3) (2022-10-05)


### Bug Fixes

* **deps:** update dependency uuid to v9 ([#9](https://github.com/croia/post-messenger/issues/9)) ([07fc66a](https://github.com/croia/post-messenger/commit/07fc66a42db84c50ea4bc0f6244a727baf83dccc))

## [1.0.2](https://github.com/croia/post-messenger/compare/v1.0.1...v1.0.2) (2022-10-04)


### Bug Fixes

* order of prepare step ([cd6eaf6](https://github.com/croia/post-messenger/commit/cd6eaf63f6271a8d9f3321d668c349953719aa07))
* try defaults for all plugins ([#12](https://github.com/croia/post-messenger/issues/12)) ([956d101](https://github.com/croia/post-messenger/commit/956d101d43f6458839404d8d7432866b1e14b05d))
* try reordering verifyConditions step ([#14](https://github.com/croia/post-messenger/issues/14)) ([212f317](https://github.com/croia/post-messenger/commit/212f317ff36df761c1d5109888e7f970f831cc51))

## [1.0.2](https://github.com/croia/post-messenger/compare/v1.0.1...v1.0.2) (2022-10-04)


### Bug Fixes

* try defaults for all plugins ([#12](https://github.com/croia/post-messenger/issues/12)) ([956d101](https://github.com/croia/post-messenger/commit/956d101d43f6458839404d8d7432866b1e14b05d))
* try reordering verifyConditions step ([#14](https://github.com/croia/post-messenger/issues/14)) ([212f317](https://github.com/croia/post-messenger/commit/212f317ff36df761c1d5109888e7f970f831cc51))

## [1.0.1](https://github.com/croia/post-messenger/compare/v1.0.0...v1.0.1) (2022-10-04)


### Bug Fixes

* try creating github release ([#11](https://github.com/croia/post-messenger/issues/11)) ([d1bfa01](https://github.com/croia/post-messenger/commit/d1bfa012f1217c99ee6e614203f37cc9593e2a00))

# 1.0.0 (2022-10-04)


### Features

* first release ([#10](https://github.com/croia/post-messenger/issues/10)) ([963cd20](https://github.com/croia/post-messenger/commit/963cd202e8c23d4294753bb6969f1b1a40fe124f))


test/mocks.ts
const cryptoKeyMock = {
  algorithm: { length: 256, name: 'AES-CBC' },
  extractable: true,
  type: 'secret',
  usages: ['encrypt', 'decrypt'],
};

export const exportedKey = {
  alg: 'A256CBC',
  ext: true,
  k: 'sIs9QIClRRbJlk8Ov1oKAOHkF_WVtApnkVYYjw-JwZM',
  key_ops: ['encrypt', 'decrypt'],
  kty: 'oct',
};

// See https://bit.ly/3eLTTjg. In >= node 15 we can assign this to window.crypto.webcrypto
Object.defineProperty(global.self, 'crypto', {
  value: {
    getRandomValues: (): Uint8Array => new Uint8Array([15, 177, 95, 199, 168, 147, 123, 246, 43, 133, 14, 112, 205, 210, 235, 113]),
    subtle: {
      decrypt: (): Promise<number[]> => Promise.resolve([32123]),
      encrypt: (): Promise<number[]> => Promise.resolve([
        55821, 22175, 31567, 37244, 33611, 45482, 30160, 23024, 41052, 59302, 26244, 56502, 15797, 48944, 7077, 40078,
      ]),
      exportKey: (): Promise<any> => Promise.resolve(exportedKey),
      generateKey: (): Promise<any> => Promise.resolve(cryptoKeyMock),
      importKey: (): Promise<any> => Promise.resolve(cryptoKeyMock),
    },
  },
});

export class TextEncoder {
  encode(): string { return 'sjdnfnsdf'; }
}

export const textDecoderResponse = '{"a":"something"}';
export class TextDecoder {
  decode(): string { return textDecoderResponse; }
}


test/factories.ts
export function buildWindow(extra?: Partial<Window>): any {
  return {
    origin: 'https://one.com',
    postMessage: () => {},
    ...extra,
  };
}

const requiredMessageEventProperties = {
  AT_TARGET: 0,
  BUBBLING_PHASE: 0,
  CAPTURING_PHASE: 0,
  NONE: 0,
  bubbles: false,
  cancelBubble: false,
  cancelable: false,
  composed: false,
  composedPath: () => [],
  defaultPrevented: false,
  eventPhase: 2,
  initEvent: () => undefined,
  initMessageEvent: () => undefined,
  isTrusted: true,
  lastEventId: '',
  ports: [],
  preventDefault: () => undefined,
  returnValue: true,
  stopImmediatePropagation: () => undefined,
  stopPropagation: () => undefined,
  timeStamp: 50000,
  type: 'message',
};

export function buildMessageEvent(extra?: Partial<WindowEventMap['message']>): WindowEventMap['message'] {
  const origin = 'https://one.com';
  const windowOne = buildWindow({ origin });
  const windowTwo = buildWindow({ origin: 'https://two.com' });
  return {
    /* properties that ts requires that we don't really care about in postMessenger: */
    ...requiredMessageEventProperties,
    currentTarget: windowTwo,
    data: {},
    origin,
    source: windowOne,
    srcElement: windowOne,
    target: windowTwo,
    ...extra,
  };
}


src/index.ts
import autoBind from 'auto-bind';
import { v4 as uuidv4 } from 'uuid';
import {
  AcceptConnectionsArgs,
  ConnectArgs,
  ConnectionDetails,
  ConnectMessage,
  EncryptionValues,
  InternalRequestNames,
  InternalRequestKeys,
  isError,
  isRequestMessage,
  Listener,
  Listeners,
  RequestName,
  PostMessengerArgs,
  RemoveAllResponders,
  RemoveListener,
  RequestOptions,
  Responders,
  ValidateOriginFn,
  ValidateRequest,
} from './types';
import {
  ab2str,
  decodeBase64,
  encodeBase64,
  hasOwnProperty,
  isUndef,
  shallowCompare,
  str2ab,
} from './utils';

const AESCBC = 'AES-CBC';

class PostMessenger<T extends Record<string, string> | undefined = undefined> {
  /* prefixed to messages and log messages to help distinguish sending/receiving side */
  clientName: string;

  connection: ConnectionDetails | null = null;

  #enableLogging: boolean;

  /* when true will encrypt messages using a key determined by self or trusted connection initiator: */
  useEncryption: (requestName?: string, throwForNoConnection?: boolean | ((error: Error) => void)) => boolean;

  /* these must be kept private, prefix with # for runtime guarentee (see https://bit.ly/31tQYou) */
  #encryptionValues: EncryptionValues = {
    algorithm: null,
    iv: null,
    requestKey: null,
  };

  #listeners: Listeners = {};

  /* maxResponseTime max time to wait in ms before canceling request */
  maxResponseTime: number;

  targetWindow: Window | null = null;

  targetOrigin: string | null = null;

  /* message names to keys representing sent and expected received request names */
  #requestNames: T | undefined;

  #validateOrigin: ValidateOriginFn | null = null;

  constructor(
    {
      clientName = 'unknown',
      enableLogging = false,
      useEncryption = true,
      maxResponseTime = 10000,
    }: PostMessengerArgs,
    requestNames?: T,
  ) {
    autoBind(this);
    this.clientName = clientName;
    this.#enableLogging = enableLogging;
    this.useEncryption = (requestName, throwForNoConnection = false): boolean => {
      const useEncryptionForMessage = useEncryption && requestName !== InternalRequestNames.postMessengerConnect;
      if (useEncryptionForMessage && !this.connection && throwForNoConnection) {
        const error = this.#buildError(`Cannot send request ${requestName}. Encryption is on but there is no connected client.`);
        if (typeof throwForNoConnection === 'function') {
          throwForNoConnection(error);
        } else {
          throw error;
        }
      }
      return useEncryptionForMessage;
    };

    if (requestNames && hasOwnProperty(requestNames, InternalRequestKeys.postMessengerConnect)) {
      throw this.#buildError(`${InternalRequestKeys.postMessengerConnect} is a reserved request name.`);
    }

    this.maxResponseTime = maxResponseTime;
    this.#requestNames = requestNames;

    this.#logger('created');
  }

  get requestNames(): T {
    return this.#requestNames as T;
  }

  #prefix(str: string): string {
    return `postMessenger: ${this.clientName} ${str}`;
  }

  #logger(...args: unknown[]): void {
    if (this.#enableLogging) {
      if (typeof args[0] === 'string') {
        console.log(this.#prefix(args[0]), ...args.slice(1));
      } else {
        console.log(...args);
      }
    }
  }

  #buildError(msg: string): Error {
    return new Error(this.#prefix(msg));
  }

  #addListener(messageName: string, fn: Listener): RemoveListener {
    if (this.#listeners[messageName]) {
      this.#listeners[messageName].push(fn);
    } else {
      this.#listeners[messageName] = [fn];
    }
    return () => this.#removeListener(messageName, fn);
  }

  #removeListener(messageName: string, fn: Listener): void {
    if (this.#listeners[messageName]) {
      const i = this.#listeners[messageName].indexOf(fn);
      if (i > -1) {
        this.#listeners[messageName].splice(i, i + 1);
      }
    }
  }

  onReceiveMessage(event: WindowEventMap['message']): void {
    if (event.data && this.#listeners[event.data.requestName]) {
      if (this.#validateOrigin) {
        if (!this.#validateOrigin(event.origin)) {
          return;
        }
      }
      this.#listeners[event.data.requestName].forEach((listener) => {
        listener(event.data, event);
      });
    }
  }

  /* Sends a single message with no expectation for a response. Internal use only, clients should use request */
  #send(message: unknown = {}): void {
    if (!this.targetWindow || !this.targetOrigin) {
      throw this.#buildError('targetWindow has not been initialized, please ensure you call setTarget before calling beginListening');
    }
    this.targetWindow.postMessage(message, this.targetOrigin);
  }

  /* Build and send RequestMessage shape expected by this.request */
  async #sendRequestMessage(requestName: string, requestId: string, messageData: unknown = {}, errorMessageStr?: string): Promise<void> {
    let data = messageData;
    let errorMessage = errorMessageStr || null;
    if (this.useEncryption(requestName, true)) {
      data = await this.#encrypt(messageData);
      if (errorMessage) {
        errorMessage = await this.#encrypt(errorMessage);
      }
    }
    this.#send({
      data,
      errorMessage,
      isError: Boolean(errorMessage),
      requestId,
      requestName,
    });
  }

  /* Sends a message and listens for a response matching a unique message id. */
  async #request<R = unknown>(requestName: string, data: unknown = {}, options: RequestOptions = {}): Promise<R> {
    const requestId = uuidv4();
    this.#logger(`sending request with name '${requestName}' to '${this.targetOrigin}':`, data);
    const result = new Promise((resolve, reject) => {
      let hasCompleted = false;
      const removeResponseListener = this.#addListener(requestName, async (responseMessage): Promise<void> => {
        if (!isRequestMessage<R>(responseMessage)) {
          return;
        }
        if (responseMessage.requestId === requestId) {
          hasCompleted = true;
          removeResponseListener();
          if (responseMessage.isError) {
            let errorMessage = responseMessage.errorMessage;
            if (this.useEncryption(requestName, true) && responseMessage.errorMessage) {
              errorMessage = await this.#decrypt(responseMessage.errorMessage);
            }
            reject(this.#buildError(
              `Responder for request name '${requestName}' to target '${this.targetOrigin}' ` +
              `failed with message: "${errorMessage}"`,
            ));
          } else {
            let responseMessageData = responseMessage.data;
            if (this.useEncryption(requestName, true)) {
              if (typeof responseMessage.data !== 'string') {
                reject(this.#buildError(
                  `encryption is required but request received a non string data response for message: ${requestName}`,
                ));
                return;
              }
              responseMessageData = await this.#decrypt<R>(responseMessage.data);
            }
            resolve(responseMessageData);
          }
        }
      });

      setTimeout(() => {
        if (!hasCompleted) {
          reject(this.#buildError(
            `Time out waiting for target '${this.targetOrigin}' to respond to request name '${requestName}'`,
          ));
          removeResponseListener();
        }
      }, options.maxResponseTime || this.maxResponseTime);
    });

    await this.#sendRequestMessage(requestName, requestId, data);

    return result as Promise<R>;
  }

  /* validate requestName exists if optional requestNames are provided to constructor */
  #getRequestName(requestName: string): string {
    if (InternalRequestNames[requestName]) {
      return InternalRequestNames[requestName];
    }

    if (this.requestNames) {
      if (!this.requestNames[requestName]) {
        throw this.#buildError(
          `requestNames were provided to constructor but unable to find requestName for ${String(requestName)}`,
        );
      }
      return this.requestNames[requestName];
    }
    return requestName;
  }

  /* type safe public request wrapper for #requestNames */
  request<R = unknown>(requestName: RequestName<T>, data: unknown = {}, options: RequestOptions = {}): Promise<R> {
    if (this.requestNames && this.connection && this.connection.requestNames && !this.connection.requestNames[String(requestName)]) {
      throw this.#buildError(
        `Connected client ${this.connection.clientName} does not have a matching request name for ${String(requestName)} so this request will fail.`,
      );
    }
    return this.#request<R>(this.#getRequestName(requestName), data, options);
  }

  /* Accepts an object of event requestNames mapping to handlers that return promises.
     Adds listeners that expect messages sent by the request function above in
     order to return a corresponding requestId */
  #bindResponders(responders: Responders<T> | Responders<InternalRequestNames>, validateRequest: ValidateRequest | null = null): RemoveAllResponders {
    const allRemoveFns: RemoveListener[] = [];
    Object.entries(responders).forEach(([messageName, handler]) => {
      const requestName = this.#getRequestName(messageName);
      const removeListenerFn = this.#addListener(requestName, async (message, event: WindowEventMap['message']): Promise<void> => {
        if (!isRequestMessage(message) || !handler) {
          return;
        }

        if (validateRequest && !validateRequest(message.data)) {
          return;
        }

        let { data } = message;
        try {
          if (this.useEncryption(requestName, true)) {
            if (typeof data !== 'string') {
              throw this.#buildError('encryption is required but responder received a non string data response');
            }
            data = await this.#decrypt(data);
          }
          const response = await handler(data, event);
          this.#logger(`responding to request name '${requestName}' from target '${this.targetOrigin}':`, response);
          this.#sendRequestMessage(requestName, message.requestId, response);
        } catch (e) {
          if (isError(e)) {
            this.#sendRequestMessage(requestName, message.requestId, {}, e.message);
          } else {
            this.#sendRequestMessage(requestName, message.requestId, {}, this.#prefix('responder threw a non Error object'));
          }
        }
      });
      allRemoveFns.push(removeListenerFn);
    });

    return () => {
      this.#logger('removing responders:', responders);
      allRemoveFns.forEach(removeFn => removeFn());
    };
  }

  bindResponders(responders: Responders<T>): RemoveAllResponders {
    if (hasOwnProperty(responders, InternalRequestKeys.postMessengerConnect)) {
      throw this.#buildError(`${InternalRequestKeys.postMessengerConnect} is a reserved request name.`);
    }
    return this.#bindResponders(responders);
  }

  async connect({ targetWindow, targetOrigin, maxRetries = 10 }: ConnectArgs): Promise<boolean> {
    if (!targetWindow || !targetOrigin) {
      throw this.#buildError('targetWindow and targetOrigin are required for connect');
    }
    this.#setTarget(targetWindow, targetOrigin);
    this.#beginListening(origin => (origin === new URL(targetOrigin).origin));

    let iv: Uint8Array | null = null;
    let jsonRequestKey: JsonWebKey | null = null;
    const useEncryption = this.useEncryption();
    if (useEncryption) {
      iv = crypto.getRandomValues(new Uint8Array(16));
      /* encryption code based examples https://bit.ly/3ex4DiQ and https://ibm.co/30ABCdZ */

      this.#encryptionValues.requestKey = await crypto.subtle.generateKey(
        { length: 256, name: AESCBC }, // AES in CBC mode, with a key length of 256 bits.
        true, // Allow extracting the key material
        ['encrypt', 'decrypt'], // Restrict usage of the key
      );
      jsonRequestKey = await crypto.subtle.exportKey('jwk', this.#encryptionValues.requestKey);

      // AES-CBC requires a 128-bit initialization vector (iv).
      this.#encryptionValues.iv = iv;

      // The algorithm is still AES-CBC. The 128-bit iv must be specified.
      this.#encryptionValues.algorithm = { iv, name: AESCBC };
    }

    const tries = maxRetries || 1;
    const maxResponseTime = 500;
    let connection: null | ConnectionDetails = null;
    for (let i = 0; i < tries; i += 1) {
      try {
        connection = await this.#request<ConnectionDetails>(InternalRequestNames.postMessengerConnect, {
          clientName: this.clientName,
          iv,
          jsonRequestKey,
          origin: window.location.origin,
          requestNames: this.requestNames,
          useEncryption,
        }, { maxResponseTime });
      } catch (e) { /* ignore */ }

      if (connection) {
        this.connection = connection;
        break;
      }
    }

    if (!this.connection) {
      throw this.#buildError(`Connection failed after ${tries} attempts over ${(tries * maxResponseTime) / 1000} seconds.`);
    }

    this.#logger(`Connection established to ${this.connection.clientName}`, this.connection);

    return true;
  }

  acceptConnections({
    allowAnyOrigin = false,
    fromClientName = null,
    origin,
  }: AcceptConnectionsArgs): Promise<ConnectionDetails> {
    if (!allowAnyOrigin && !origin) {
      throw this.#buildError('allowAnyOrigin must be true if origin is not specified');
    }

    /* Optionally allow caller to specify client name to accept connections from. This helps
       disambiguate multiple postMessenger connections on the same page to / from same origin */
    const validateConnectRequest = (data: ConnectMessage) => (
      !fromClientName || fromClientName === data.clientName
    );

    this.#beginListening(messageOrigin => (origin ? messageOrigin === origin : true));

    return new Promise((resolve, reject) => {
      const removeConnectionResponder = this.#bindResponders({
        postMessengerConnect: async (data: ConnectMessage, event): Promise<ConnectionDetails> => {
          if (!event.source) {
            reject(this.#buildError('Received connection attempt but event.source is null'));
          }

          this.connection = {
            clientName: data.clientName,
            requestNames: data.requestNames,
            useEncryption: data.useEncryption,
          };

          this.#setTarget(event.source as Window, data.origin);

          const requestNames = this.requestNames;
          /* check to make sure both sides agree on requestNames */
          const providedRequestNamesObjMatch = (
            !!requestNames &&
            !!data.requestNames &&
            shallowCompare(requestNames, data.requestNames)
          );

          /* either they are both provided and match or both should be undefined: */
          if (!providedRequestNamesObjMatch && !(isUndef(this.requestNames) && isUndef(data.requestNames))) {
            reject(this.#buildError('Received connection attempt but requestNames must be the same for both PostMessenger instances'));
          }

          /* check to make sure both sides agree on useEncryption */
          const useEncryption = this.useEncryption();
          if (useEncryption !== data.useEncryption) {
            reject(this.#buildError('Received connection attempt but useEncryption must be the same for both PostMessenger instances'));
          }

          if (this.useEncryption()) {
            if (!data.iv || !data.jsonRequestKey || !data.useEncryption) {
              const error = this.#buildError(
                'encryption is required but iv or jsonRequestKey or useEncryption were not provided in connection message.',
              );
              reject(error);
              throw error;
            }

            this.#encryptionValues.iv = new Uint8Array([...data.iv]);
            this.#encryptionValues.algorithm = { iv: this.#encryptionValues.iv, name: AESCBC };
            this.#encryptionValues.requestKey = await crypto.subtle.importKey(
              'jwk',
              data.jsonRequestKey,
              { name: AESCBC },
              false,
              ['encrypt', 'decrypt'],
            );
          }

          /* If the page has multiple postMessenger instances to / from same origin there could be
         a second connect message that overwrites this connection. Once the first connection
         is confirmed we should stop listening for future connect messages: */
          removeConnectionResponder();

          this.#logger(`Accepted connection from ${this.connection.clientName}`, this.connection);

          resolve(this.connection);

          return {
            clientName: this.clientName,
            requestNames: data.requestNames,
            useEncryption: this.useEncryption(),
          };
        },
      }, validateConnectRequest);
    });
  }

  #setTarget(targetWindow: Window, targetOrigin: string): void {
    if (!targetWindow || !targetOrigin) {
      throw this.#buildError('targetWindow and targetWindow are required for setTarget');
    }
    this.targetWindow = targetWindow;
    /* validate initTargetOrigin is proper URL: */
    const targetUrl = new URL(targetOrigin);
    this.targetOrigin = targetUrl.origin;
  }

  #beginListening(validateOrigin: ValidateOriginFn): void {
    this.#validateOrigin = validateOrigin;
    window.addEventListener('message', this.onReceiveMessage);
  }

  stopListening(): void {
    window.removeEventListener('message', this.onReceiveMessage);
  }

  async #decrypt<R = unknown>(data: string): Promise<R> {
    if (!this.#encryptionValues.algorithm || !this.#encryptionValues.requestKey) {
      throw this.#buildError('encryptionValues must be set before calling decrpyt');
    }

    const base64Decoded = decodeBase64(data);
    const encodedData = str2ab(base64Decoded);
    const decryptedData = await crypto.subtle.decrypt(
      this.#encryptionValues.algorithm,
      this.#encryptionValues.requestKey,
      encodedData,
    );

    if (decryptedData.byteLength === 0) {
      return null as unknown as R;
    }

    const textDecoder = new TextDecoder();
    const decryptedText = textDecoder.decode(decryptedData);
    return JSON.parse(decryptedText);
  }

  /* encryption code based examples https://bit.ly/3ex4DiQ and https://ibm.co/30ABCdZ */
  async #encrypt(data: unknown): Promise<string> {
    if (!this.#encryptionValues.algorithm || !this.#encryptionValues.requestKey) {
      throw this.#buildError('encryptionValues must be set before calling encrypt');
    }

    // This is the plaintext:
    const encoder = new TextEncoder();
    const encodedData = encoder.encode(JSON.stringify(data));

    // Finally, encrypt the plaintext, and obtain the ciphertext.
    const encryptedAB = await crypto.subtle.encrypt(
      this.#encryptionValues.algorithm,
      // This must be an AES-CBC key encryption key, or this fn will reject.
      this.#encryptionValues.requestKey,
      // The plaintext to encrypt.
      encodedData,
    );

    const encryptedText = ab2str(encryptedAB);
    const base64Text = encodeBase64(encryptedText);
    return base64Text;
  }
}

export { PostMessenger };
